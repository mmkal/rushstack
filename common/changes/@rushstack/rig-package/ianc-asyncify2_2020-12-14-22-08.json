{
  "changes": [
    {
      "packageName": "@rushstack/rig-package",
      "comment": "",
      "type": "none"
    }
  ],
  "packageName": "@rushstack/rig-package",
  "email": "iclanton@users.noreply.github.com"
}