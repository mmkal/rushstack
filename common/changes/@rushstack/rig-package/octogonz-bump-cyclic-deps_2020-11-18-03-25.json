{
  "changes": [
    {
      "packageName": "@rushstack/rig-package",
      "comment": "",
      "type": "none"
    }
  ],
  "packageName": "@rushstack/rig-package",
  "email": "4673363+octogonz@users.noreply.github.com"
}