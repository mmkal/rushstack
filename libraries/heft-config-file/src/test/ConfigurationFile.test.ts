// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import * as nodeJsPath from 'path';

import { ConfigurationFile, PathResolutionMethod, InheritanceType } from '../ConfigurationFile';
import {
  FileSystem,
  JsonFile,
  Path,
  StringBufferTerminalProvider,
  Terminal
} from '@rushstack/node-core-library';
import { RigConfig } from '@rushstack/rig-package';

describe('ConfigurationFile', () => {
  const projectRoot: string = nodeJsPath.resolve(__dirname, '..', '..');
  let terminalProvider: StringBufferTerminalProvider;
  let terminal: Terminal;

  beforeEach(() => {
    const projectRoot: string = nodeJsPath.resolve(__dirname, '..', '..');
    const formatPathForLogging: (path: string) => string = (path: string) =>
      `<project root>/${Path.convertToSlashes(nodeJsPath.relative(projectRoot, path))}`;
    jest.spyOn(ConfigurationFile, '_formatPathForLogging').mockImplementation(formatPathForLogging);
    jest.spyOn(JsonFile, '_formatPathForError').mockImplementation(formatPathForLogging);

    terminalProvider = new StringBufferTerminalProvider(false);
    terminal = new Terminal(terminalProvider);
  });

  afterEach(() => {
    expect({
      log: terminalProvider.getOutput(),
      warning: terminalProvider.getWarningOutput(),
      error: terminalProvider.getErrorOutput(),
      verbose: terminalProvider.getVerbose()
    }).toMatchSnapshot();
  });

  describe('A simple config file', () => {
    const configFileFolderName: string = 'simplestConfigFile';
    const projectRelativeFilePath: string = `${configFileFolderName}/simplestConfigFile.json`;
    const schemaPath: string = nodeJsPath.resolve(
      __dirname,
      configFileFolderName,
      'simplestConfigFile.schema.json'
    );

    interface ISimplestConfigFile {
      thing: string;
    }

    it('Correctly loads the config file', async () => {
      const configFileLoader: ConfigurationFile<ISimplestConfigFile> = new ConfigurationFile<
        ISimplestConfigFile
      >({ projectRelativeFilePath: projectRelativeFilePath, jsonSchemaPath: schemaPath });
      const loadedConfigFile: ISimplestConfigFile = await configFileLoader.loadConfigurationFileForProjectAsync(
        terminal,
        __dirname
      );
      const expectedConfigFile: ISimplestConfigFile = { thing: 'A' };

      expect(JSON.stringify(loadedConfigFile)).toEqual(JSON.stringify(expectedConfigFile));
      expect(configFileLoader.getObjectSourceFilePath(loadedConfigFile)).toEqual(
        nodeJsPath.resolve(__dirname, projectRelativeFilePath)
      );
      expect(
        configFileLoader.getPropertyOriginalValue({ parentObject: loadedConfigFile, propertyName: 'thing' })
      ).toEqual('A');
    });

    it('Correctly resolves paths relative to the config file', async () => {
      const configFileLoader: ConfigurationFile<ISimplestConfigFile> = new ConfigurationFile<
        ISimplestConfigFile
      >({
        projectRelativeFilePath: projectRelativeFilePath,
        jsonSchemaPath: schemaPath,
        jsonPathMetadata: {
          '$.thing': {
            pathResolutionMethod: PathResolutionMethod.resolvePathRelativeToConfigurationFile
          }
        }
      });
      const loadedConfigFile: ISimplestConfigFile = await configFileLoader.loadConfigurationFileForProjectAsync(
        terminal,
        __dirname
      );
      const expectedConfigFile: ISimplestConfigFile = {
        thing: nodeJsPath.resolve(__dirname, configFileFolderName, 'A')
      };
      expect(JSON.stringify(loadedConfigFile)).toEqual(JSON.stringify(expectedConfigFile));
      expect(configFileLoader.getObjectSourceFilePath(loadedConfigFile)).toEqual(
        nodeJsPath.resolve(__dirname, projectRelativeFilePath)
      );
      expect(
        configFileLoader.getPropertyOriginalValue({ parentObject: loadedConfigFile, propertyName: 'thing' })
      ).toEqual('A');
    });

    it('Correctly resolves paths relative to the project root', async () => {
      const configFileLoader: ConfigurationFile<ISimplestConfigFile> = new ConfigurationFile<
        ISimplestConfigFile
      >({
        projectRelativeFilePath: projectRelativeFilePath,
        jsonSchemaPath: schemaPath,
        jsonPathMetadata: {
          '$.thing': {
            pathResolutionMethod: PathResolutionMethod.resolvePathRelativeToProjectRoot
          }
        }
      });
      const loadedConfigFile: ISimplestConfigFile = await configFileLoader.loadConfigurationFileForProjectAsync(
        terminal,
        __dirname
      );
      const expectedConfigFile: ISimplestConfigFile = {
        thing: nodeJsPath.resolve(projectRoot, 'A')
      };
      expect(JSON.stringify(loadedConfigFile)).toEqual(JSON.stringify(expectedConfigFile));
      expect(configFileLoader.getObjectSourceFilePath(loadedConfigFile)).toEqual(
        nodeJsPath.resolve(__dirname, projectRelativeFilePath)
      );
      expect(
        configFileLoader.getPropertyOriginalValue({ parentObject: loadedConfigFile, propertyName: 'thing' })
      ).toEqual('A');
    });
  });

  describe('A simple config file containing an array', () => {
    const configFileFolderName: string = 'simpleConfigFile';
    const projectRelativeFilePath: string = `${configFileFolderName}/simpleConfigFile.json`;
    const schemaPath: string = nodeJsPath.resolve(
      __dirname,
      configFileFolderName,
      'simpleConfigFile.schema.json'
    );

    interface ISimpleConfigFile {
      things: string[];
    }

    it('Correctly loads the config file', async () => {
      const configFileLoader: ConfigurationFile<ISimpleConfigFile> = new ConfigurationFile<ISimpleConfigFile>(
        { projectRelativeFilePath: projectRelativeFilePath, jsonSchemaPath: schemaPath }
      );
      const loadedConfigFile: ISimpleConfigFile = await configFileLoader.loadConfigurationFileForProjectAsync(
        terminal,
        __dirname
      );
      const expectedConfigFile: ISimpleConfigFile = { things: ['A', 'B', 'C'] };
      expect(JSON.stringify(loadedConfigFile)).toEqual(JSON.stringify(expectedConfigFile));
    });

    it('Correctly resolves paths relative to the config file', async () => {
      const configFileLoader: ConfigurationFile<ISimpleConfigFile> = new ConfigurationFile<ISimpleConfigFile>(
        {
          projectRelativeFilePath: projectRelativeFilePath,
          jsonSchemaPath: schemaPath,
          jsonPathMetadata: {
            '$.things.*': {
              pathResolutionMethod: PathResolutionMethod.resolvePathRelativeToConfigurationFile
            }
          }
        }
      );
      const loadedConfigFile: ISimpleConfigFile = await configFileLoader.loadConfigurationFileForProjectAsync(
        terminal,
        __dirname
      );
      const expectedConfigFile: ISimpleConfigFile = {
        things: [
          nodeJsPath.resolve(__dirname, configFileFolderName, 'A'),
          nodeJsPath.resolve(__dirname, configFileFolderName, 'B'),
          nodeJsPath.resolve(__dirname, configFileFolderName, 'C')
        ]
      };
      expect(JSON.stringify(loadedConfigFile)).toEqual(JSON.stringify(expectedConfigFile));
    });

    it('Correctly resolves paths relative to the project root', async () => {
      const configFileLoader: ConfigurationFile<ISimpleConfigFile> = new ConfigurationFile<ISimpleConfigFile>(
        {
          projectRelativeFilePath: projectRelativeFilePath,
          jsonSchemaPath: schemaPath,
          jsonPathMetadata: {
            '$.things.*': {
              pathResolutionMethod: PathResolutionMethod.resolvePathRelativeToProjectRoot
            }
          }
        }
      );
      const loadedConfigFile: ISimpleConfigFile = await configFileLoader.loadConfigurationFileForProjectAsync(
        terminal,
        __dirname
      );
      const expectedConfigFile: ISimpleConfigFile = {
        things: [
          nodeJsPath.resolve(projectRoot, 'A'),
          nodeJsPath.resolve(projectRoot, 'B'),
          nodeJsPath.resolve(projectRoot, 'C')
        ]
      };
      expect(JSON.stringify(loadedConfigFile)).toEqual(JSON.stringify(expectedConfigFile));
    });
  });

  describe('A simple config file with "extends"', () => {
    const configFileFolderName: string = 'simpleConfigFileWithExtends';
    const projectRelativeFilePath: string = `${configFileFolderName}/simpleConfigFileWithExtends.json`;
    const schemaPath: string = nodeJsPath.resolve(
      __dirname,
      configFileFolderName,
      'simpleConfigFileWithExtends.schema.json'
    );

    interface ISimpleConfigFile {
      things: string[];
    }

    it('Correctly loads the config file with default config meta', async () => {
      const configFileLoader: ConfigurationFile<ISimpleConfigFile> = new ConfigurationFile<ISimpleConfigFile>(
        { projectRelativeFilePath: projectRelativeFilePath, jsonSchemaPath: schemaPath }
      );
      const loadedConfigFile: ISimpleConfigFile = await configFileLoader.loadConfigurationFileForProjectAsync(
        terminal,
        __dirname
      );
      const expectedConfigFile: ISimpleConfigFile = { things: ['A', 'B', 'C', 'D', 'E'] };
      expect(JSON.stringify(loadedConfigFile)).toEqual(JSON.stringify(expectedConfigFile));
    });

    it('Correctly loads the config file with "append" in config meta', async () => {
      const configFileLoader: ConfigurationFile<ISimpleConfigFile> = new ConfigurationFile<ISimpleConfigFile>(
        {
          projectRelativeFilePath: projectRelativeFilePath,
          jsonSchemaPath: schemaPath,
          propertyInheritance: {
            things: {
              inheritanceType: InheritanceType.append
            }
          }
        }
      );
      const loadedConfigFile: ISimpleConfigFile = await configFileLoader.loadConfigurationFileForProjectAsync(
        terminal,
        __dirname
      );
      const expectedConfigFile: ISimpleConfigFile = { things: ['A', 'B', 'C', 'D', 'E'] };
      expect(JSON.stringify(loadedConfigFile)).toEqual(JSON.stringify(expectedConfigFile));
    });

    it('Correctly loads the config file with "replace" in config meta', async () => {
      const configFileLoader: ConfigurationFile<ISimpleConfigFile> = new ConfigurationFile<ISimpleConfigFile>(
        {
          projectRelativeFilePath: projectRelativeFilePath,
          jsonSchemaPath: schemaPath,
          propertyInheritance: {
            things: {
              inheritanceType: InheritanceType.replace
            }
          }
        }
      );
      const loadedConfigFile: ISimpleConfigFile = await configFileLoader.loadConfigurationFileForProjectAsync(
        terminal,
        __dirname
      );
      const expectedConfigFile: ISimpleConfigFile = { things: ['D', 'E'] };
      expect(JSON.stringify(loadedConfigFile)).toEqual(JSON.stringify(expectedConfigFile));
    });

    it('Correctly loads the config file with "custom" in config meta', async () => {
      const configFileLoader: ConfigurationFile<ISimpleConfigFile> = new ConfigurationFile<ISimpleConfigFile>(
        {
          projectRelativeFilePath: projectRelativeFilePath,
          jsonSchemaPath: schemaPath,
          propertyInheritance: {
            things: {
              inheritanceType: InheritanceType.custom,
              inheritanceFunction: (current: string[], parent: string[]) => ['X', 'Y', 'Z']
            }
          }
        }
      );
      const loadedConfigFile: ISimpleConfigFile = await configFileLoader.loadConfigurationFileForProjectAsync(
        terminal,
        __dirname
      );
      const expectedConfigFile: ISimpleConfigFile = { things: ['X', 'Y', 'Z'] };
      expect(JSON.stringify(loadedConfigFile)).toEqual(JSON.stringify(expectedConfigFile));
    });

    it('Correctly resolves paths relative to the config file', async () => {
      const configFileLoader: ConfigurationFile<ISimpleConfigFile> = new ConfigurationFile<ISimpleConfigFile>(
        {
          projectRelativeFilePath: projectRelativeFilePath,
          jsonSchemaPath: schemaPath,
          jsonPathMetadata: {
            '$.things.*': {
              pathResolutionMethod: PathResolutionMethod.resolvePathRelativeToConfigurationFile
            }
          }
        }
      );
      const loadedConfigFile: ISimpleConfigFile = await configFileLoader.loadConfigurationFileForProjectAsync(
        terminal,
        __dirname
      );
      const parentConfigFileFolder: string = nodeJsPath.resolve(
        __dirname,
        configFileFolderName,
        '..',
        'simpleConfigFile'
      );

      const expectedConfigFile: ISimpleConfigFile = {
        things: [
          nodeJsPath.resolve(parentConfigFileFolder, 'A'),
          nodeJsPath.resolve(parentConfigFileFolder, 'B'),
          nodeJsPath.resolve(parentConfigFileFolder, 'C'),
          nodeJsPath.resolve(__dirname, configFileFolderName, 'D'),
          nodeJsPath.resolve(__dirname, configFileFolderName, 'E')
        ]
      };
      expect(JSON.stringify(loadedConfigFile)).toEqual(JSON.stringify(expectedConfigFile));
    });
  });

  describe('A complex config file', () => {
    interface IComplexConfigFile {
      plugins: { plugin: string }[];
    }

    it('Correctly loads a complex config file', async () => {
      const projectRelativeFilePath: string = 'complexConfigFile/pluginsB.json';
      const parentConfigFilePath: string = nodeJsPath.resolve(
        __dirname,
        'complexConfigFile',
        'pluginsA.json'
      );
      const schemaPath: string = nodeJsPath.resolve(__dirname, 'complexConfigFile', 'plugins.schema.json');

      const configFileLoader: ConfigurationFile<IComplexConfigFile> = new ConfigurationFile<
        IComplexConfigFile
      >({
        projectRelativeFilePath: projectRelativeFilePath,
        jsonSchemaPath: schemaPath,
        jsonPathMetadata: {
          '$.plugins.*.plugin': {
            pathResolutionMethod: PathResolutionMethod.NodeResolve
          }
        }
      });
      const loadedConfigFile: IComplexConfigFile = await configFileLoader.loadConfigurationFileForProjectAsync(
        terminal,
        __dirname
      );
      const expectedConfigFile: IComplexConfigFile = {
        plugins: [
          {
            plugin: await FileSystem.getRealPathAsync(
              nodeJsPath.resolve(
                projectRoot,
                'node_modules',
                '@rushstack',
                'node-core-library',
                'lib',
                'index.js'
              )
            )
          },
          {
            plugin: await FileSystem.getRealPathAsync(
              nodeJsPath.resolve(projectRoot, 'node_modules', '@rushstack', 'heft', 'lib', 'index.js')
            )
          },
          {
            plugin: await FileSystem.getRealPathAsync(
              nodeJsPath.resolve(projectRoot, 'node_modules', '@rushstack', 'eslint-config', 'index.js')
            )
          }
        ]
      };

      expect(JSON.stringify(loadedConfigFile)).toEqual(JSON.stringify(expectedConfigFile));

      expect(
        configFileLoader.getPropertyOriginalValue({
          parentObject: loadedConfigFile.plugins[0],
          propertyName: 'plugin'
        })
      ).toEqual('@rushstack/node-core-library');
      expect(
        configFileLoader.getPropertyOriginalValue({
          parentObject: loadedConfigFile.plugins[1],
          propertyName: 'plugin'
        })
      ).toEqual('@rushstack/heft');
      expect(
        configFileLoader.getPropertyOriginalValue({
          parentObject: loadedConfigFile.plugins[2],
          propertyName: 'plugin'
        })
      ).toEqual('@rushstack/eslint-config');

      expect(configFileLoader.getObjectSourceFilePath(loadedConfigFile.plugins[0])).toEqual(
        parentConfigFilePath
      );
      expect(configFileLoader.getObjectSourceFilePath(loadedConfigFile.plugins[1])).toEqual(
        nodeJsPath.resolve(__dirname, projectRelativeFilePath)
      );
      expect(configFileLoader.getObjectSourceFilePath(loadedConfigFile.plugins[2])).toEqual(
        nodeJsPath.resolve(__dirname, projectRelativeFilePath)
      );
    });
  });

  describe('loading a rig', () => {
    const projectFolder: string = nodeJsPath.resolve(__dirname, 'project-referencing-rig');
    const rigConfig: RigConfig = RigConfig.loadForProjectFolder({ projectFolderPath: projectFolder });

    const schemaPath: string = nodeJsPath.resolve(
      __dirname,
      'simplestConfigFile',
      'simplestConfigFile.schema.json'
    );

    interface ISimplestConfigFile {
      thing: string;
    }

    it('correctly loads a config file inside a rig', async () => {
      const projectRelativeFilePath: string = 'config/simplestConfigFile.json';
      const configFileLoader: ConfigurationFile<ISimplestConfigFile> = new ConfigurationFile<
        ISimplestConfigFile
      >({ projectRelativeFilePath: projectRelativeFilePath, jsonSchemaPath: schemaPath });
      const loadedConfigFile: ISimplestConfigFile = await configFileLoader.loadConfigurationFileForProjectAsync(
        terminal,
        projectFolder,
        rigConfig
      );
      const expectedConfigFile: ISimplestConfigFile = { thing: 'A' };

      expect(JSON.stringify(loadedConfigFile)).toEqual(JSON.stringify(expectedConfigFile));
      expect(configFileLoader.getObjectSourceFilePath(loadedConfigFile)).toEqual(
        nodeJsPath.resolve(
          projectFolder,
          'node_modules',
          'test-rig',
          'profiles',
          'default',
          projectRelativeFilePath
        )
      );
      expect(
        configFileLoader.getPropertyOriginalValue({ parentObject: loadedConfigFile, propertyName: 'thing' })
      ).toEqual('A');
    });

    it('correctly loads a config file inside a rig via tryLoadConfigurationFileForProjectAsync', async () => {
      const projectRelativeFilePath: string = 'config/simplestConfigFile.json';
      const configFileLoader: ConfigurationFile<ISimplestConfigFile> = new ConfigurationFile<
        ISimplestConfigFile
      >({ projectRelativeFilePath: projectRelativeFilePath, jsonSchemaPath: schemaPath });
      const loadedConfigFile:
        | ISimplestConfigFile
        | undefined = await configFileLoader.tryLoadConfigurationFileForProjectAsync(
        terminal,
        projectFolder,
        rigConfig
      );
      const expectedConfigFile: ISimplestConfigFile = { thing: 'A' };

      expect(loadedConfigFile).not.toBeUndefined();
      expect(JSON.stringify(loadedConfigFile)).toEqual(JSON.stringify(expectedConfigFile));
      expect(configFileLoader.getObjectSourceFilePath(loadedConfigFile!)).toEqual(
        nodeJsPath.resolve(
          projectFolder,
          'node_modules',
          'test-rig',
          'profiles',
          'default',
          projectRelativeFilePath
        )
      );
      expect(
        configFileLoader.getPropertyOriginalValue({ parentObject: loadedConfigFile!, propertyName: 'thing' })
      ).toEqual('A');
    });

    it("throws an error when a config file doesn't exist in a project referencing a rig, which also doesn't have the file", async () => {
      const configFileLoader: ConfigurationFile<void> = new ConfigurationFile({
        projectRelativeFilePath: 'config/notExist.json',
        jsonSchemaPath: schemaPath
      });
      try {
        await configFileLoader.loadConfigurationFileForProjectAsync(terminal, projectFolder, rigConfig);
        fail();
      } catch (e) {
        expect(e).toMatchSnapshot();
      }
    });
  });

  describe('error cases', () => {
    const errorCasesFolderName: string = 'errorCases';

    it("throws an error when the file doesn't exist", async () => {
      const errorCaseFolderName: string = 'invalidType';
      const configFileLoader: ConfigurationFile<void> = new ConfigurationFile({
        projectRelativeFilePath: `${errorCasesFolderName}/${errorCaseFolderName}/notExist.json`,
        jsonSchemaPath: nodeJsPath.resolve(
          __dirname,
          errorCasesFolderName,
          errorCaseFolderName,
          'config.schema.json'
        )
      });
      try {
        await configFileLoader.loadConfigurationFileForProjectAsync(terminal, __dirname);
        fail();
      } catch (e) {
        expect(e).toMatchSnapshot();
      }
    });

    it("returns undefined when the file doesn't exist for tryLoadConfigurationFileForProjectAsync", async () => {
      const errorCaseFolderName: string = 'invalidType';
      const configFileLoader: ConfigurationFile<void> = new ConfigurationFile({
        projectRelativeFilePath: `${errorCasesFolderName}/${errorCaseFolderName}/notExist.json`,
        jsonSchemaPath: nodeJsPath.resolve(
          __dirname,
          errorCasesFolderName,
          errorCaseFolderName,
          'config.schema.json'
        )
      });
      try {
        expect(
          await configFileLoader.tryLoadConfigurationFileForProjectAsync(terminal, __dirname)
        ).toBeUndefined();
      } catch (e) {
        fail();
      }
    });

    it("Throws an error when the file isn't valid JSON", async () => {
      const errorCaseFolderName: string = 'invalidJson';
      const configFileLoader: ConfigurationFile<void> = new ConfigurationFile({
        projectRelativeFilePath: `${errorCasesFolderName}/${errorCaseFolderName}/config.json`,
        jsonSchemaPath: nodeJsPath.resolve(
          __dirname,
          errorCasesFolderName,
          errorCaseFolderName,
          'config.schema.json'
        )
      });
      try {
        await configFileLoader.loadConfigurationFileForProjectAsync(terminal, __dirname);
        fail();
      } catch (e) {
        expect(e).toMatchSnapshot();
      }
    });

    it("Throws an error for a file that doesn't match its schema", async () => {
      const errorCaseFolderName: string = 'invalidType';
      const configFileLoader: ConfigurationFile<void> = new ConfigurationFile({
        projectRelativeFilePath: `${errorCasesFolderName}/${errorCaseFolderName}/config.json`,
        jsonSchemaPath: nodeJsPath.resolve(
          __dirname,
          errorCasesFolderName,
          errorCaseFolderName,
          'config.schema.json'
        )
      });
      try {
        await configFileLoader.loadConfigurationFileForProjectAsync(terminal, __dirname);
        fail();
      } catch (e) {
        expect(e).toMatchSnapshot();
      }
    });

    it('Throws an error when there is a circular reference in "extends" properties', async () => {
      const errorCaseFolderName: string = 'circularReference';
      const configFileLoader: ConfigurationFile<void> = new ConfigurationFile({
        projectRelativeFilePath: `${errorCasesFolderName}/${errorCaseFolderName}/config1.json`,
        jsonSchemaPath: nodeJsPath.resolve(
          __dirname,
          errorCasesFolderName,
          errorCaseFolderName,
          'config.schema.json'
        )
      });
      try {
        await configFileLoader.loadConfigurationFileForProjectAsync(terminal, __dirname);
        fail();
      } catch (e) {
        expect(e).toMatchSnapshot();
      }
    });

    it('Throws an error when an "extends" property points to a file that cannot be resolved', async () => {
      const errorCaseFolderName: string = 'extendsNotExist';
      const configFileLoader: ConfigurationFile<void> = new ConfigurationFile({
        projectRelativeFilePath: `${errorCasesFolderName}/${errorCaseFolderName}/config.json`,
        jsonSchemaPath: nodeJsPath.resolve(
          __dirname,
          errorCasesFolderName,
          errorCaseFolderName,
          'config.schema.json'
        )
      });
      try {
        await configFileLoader.loadConfigurationFileForProjectAsync(terminal, __dirname);
        fail();
      } catch (e) {
        expect(e).toMatchSnapshot();
      }
    });

    it("Throws an error when a combined config file doesn't match the schema", async () => {
      const errorCaseFolderName: string = 'invalidCombinedFile';
      const configFileLoader: ConfigurationFile<void> = new ConfigurationFile({
        projectRelativeFilePath: `${errorCasesFolderName}/${errorCaseFolderName}/config1.json`,
        jsonSchemaPath: nodeJsPath.resolve(
          __dirname,
          errorCasesFolderName,
          errorCaseFolderName,
          'config.schema.json'
        )
      });

      try {
        await configFileLoader.loadConfigurationFileForProjectAsync(terminal, __dirname);
        fail();
      } catch (e) {
        expect(e).toMatchSnapshot();
      }
    });

    it("Throws an error when a requested file doesn't exist", async () => {
      const configFileLoader: ConfigurationFile<void> = new ConfigurationFile({
        projectRelativeFilePath: `${errorCasesFolderName}/folderThatDoesntExist/config.json`,
        jsonSchemaPath: nodeJsPath.resolve(
          __dirname,
          errorCasesFolderName,
          'invalidCombinedFile',
          'config.schema.json'
        )
      });

      try {
        await configFileLoader.loadConfigurationFileForProjectAsync(terminal, __dirname);
        fail();
      } catch (e) {
        expect(e).toMatchSnapshot();
      }
    });
  });
});
