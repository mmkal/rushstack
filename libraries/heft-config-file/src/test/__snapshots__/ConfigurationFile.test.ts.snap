// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfigurationFile A complex config file Correctly loads a complex config file 1`] = `
Object {
  "error": "",
  "log": "",
  "verbose": "",
  "warning": "",
}
`;

exports[`ConfigurationFile A simple config file Correctly loads the config file 1`] = `
Object {
  "error": "",
  "log": "",
  "verbose": "",
  "warning": "",
}
`;

exports[`ConfigurationFile A simple config file Correctly resolves paths relative to the config file 1`] = `
Object {
  "error": "",
  "log": "",
  "verbose": "",
  "warning": "",
}
`;

exports[`ConfigurationFile A simple config file Correctly resolves paths relative to the project root 1`] = `
Object {
  "error": "",
  "log": "",
  "verbose": "",
  "warning": "",
}
`;

exports[`ConfigurationFile A simple config file containing an array Correctly loads the config file 1`] = `
Object {
  "error": "",
  "log": "",
  "verbose": "",
  "warning": "",
}
`;

exports[`ConfigurationFile A simple config file containing an array Correctly resolves paths relative to the config file 1`] = `
Object {
  "error": "",
  "log": "",
  "verbose": "",
  "warning": "",
}
`;

exports[`ConfigurationFile A simple config file containing an array Correctly resolves paths relative to the project root 1`] = `
Object {
  "error": "",
  "log": "",
  "verbose": "",
  "warning": "",
}
`;

exports[`ConfigurationFile A simple config file with "extends" Correctly loads the config file with "append" in config meta 1`] = `
Object {
  "error": "",
  "log": "",
  "verbose": "",
  "warning": "",
}
`;

exports[`ConfigurationFile A simple config file with "extends" Correctly loads the config file with "custom" in config meta 1`] = `
Object {
  "error": "",
  "log": "",
  "verbose": "",
  "warning": "",
}
`;

exports[`ConfigurationFile A simple config file with "extends" Correctly loads the config file with "replace" in config meta 1`] = `
Object {
  "error": "",
  "log": "",
  "verbose": "",
  "warning": "",
}
`;

exports[`ConfigurationFile A simple config file with "extends" Correctly loads the config file with default config meta 1`] = `
Object {
  "error": "",
  "log": "",
  "verbose": "",
  "warning": "",
}
`;

exports[`ConfigurationFile A simple config file with "extends" Correctly resolves paths relative to the config file 1`] = `
Object {
  "error": "",
  "log": "",
  "verbose": "",
  "warning": "",
}
`;

exports[`ConfigurationFile error cases Throws an error for a file that doesn't match its schema 1`] = `
[Error: JSON validation failed:
<project root>/src/test/errorCases/invalidType/config.json

Error: #/filePaths
       Expected type array but found type string]
`;

exports[`ConfigurationFile error cases Throws an error for a file that doesn't match its schema 2`] = `
Object {
  "error": "",
  "log": "",
  "verbose": "",
  "warning": "",
}
`;

exports[`ConfigurationFile error cases Throws an error when a combined config file doesn't match the schema 1`] = `
[Error: Resolved configuration object does not match schema: Error: JSON validation failed:
<project root>/src/test/errorCases/invalidCombinedFile/config1.json

Error: #/
       Data does not match any schemas from 'oneOf'
  Error: #/
         Additional properties not allowed: folderPaths
  Error: #/
         Additional properties not allowed: filePaths]
`;

exports[`ConfigurationFile error cases Throws an error when a combined config file doesn't match the schema 2`] = `
Object {
  "error": "",
  "log": "",
  "verbose": "",
  "warning": "",
}
`;

exports[`ConfigurationFile error cases Throws an error when a requested file doesn't exist 1`] = `[Error: File does not exist: <project root>/src/test/errorCases/folderThatDoesntExist/config.json]`;

exports[`ConfigurationFile error cases Throws an error when a requested file doesn't exist 2`] = `
Object {
  "error": "",
  "log": "",
  "verbose": "Configuration file \\"<project root>/src/test/errorCases/folderThatDoesntExist/config.json\\" not found.[n]",
  "warning": "",
}
`;

exports[`ConfigurationFile error cases Throws an error when an "extends" property points to a file that cannot be resolved 1`] = `[Error: In file "<project root>/src/test/errorCases/extendsNotExist/config.json", file referenced in "extends" property ("./config2.json") cannot be resolved.]`;

exports[`ConfigurationFile error cases Throws an error when an "extends" property points to a file that cannot be resolved 2`] = `
Object {
  "error": "",
  "log": "",
  "verbose": "Configuration file \\"<project root>/src/test/errorCases/extendsNotExist/config2.json\\" not found.[n]",
  "warning": "",
}
`;

exports[`ConfigurationFile error cases Throws an error when the file isn't valid JSON 1`] = `
[Error: In config file "<project root>/src/test/errorCases/invalidJson/config.json": SyntaxError: Unexpected token '\\n' at 2:19
  "filePaths": "A
                 ^]
`;

exports[`ConfigurationFile error cases Throws an error when the file isn't valid JSON 2`] = `
Object {
  "error": "",
  "log": "",
  "verbose": "",
  "warning": "",
}
`;

exports[`ConfigurationFile error cases Throws an error when there is a circular reference in "extends" properties 1`] = `[Error: A loop has been detected in the "extends" properties of configuration file at "<project root>/src/test/errorCases/circularReference/config1.json".]`;

exports[`ConfigurationFile error cases Throws an error when there is a circular reference in "extends" properties 2`] = `
Object {
  "error": "",
  "log": "",
  "verbose": "",
  "warning": "",
}
`;

exports[`ConfigurationFile error cases returns undefined when the file doesn't exist for tryLoadConfigurationFileForProjectAsync 1`] = `
Object {
  "error": "",
  "log": "",
  "verbose": "",
  "warning": "",
}
`;

exports[`ConfigurationFile error cases throws an error when the file doesn't exist 1`] = `[Error: File does not exist: <project root>/src/test/errorCases/invalidType/notExist.json]`;

exports[`ConfigurationFile error cases throws an error when the file doesn't exist 2`] = `
Object {
  "error": "",
  "log": "",
  "verbose": "Configuration file \\"<project root>/src/test/errorCases/invalidType/notExist.json\\" not found.[n]",
  "warning": "",
}
`;

exports[`ConfigurationFile loading a rig correctly loads a config file inside a rig 1`] = `
Object {
  "error": "",
  "log": "",
  "verbose": "Configuration file \\"<project root>/src/test/project-referencing-rig/config/simplestConfigFile.json\\" not found.[n]",
  "warning": "",
}
`;

exports[`ConfigurationFile loading a rig correctly loads a config file inside a rig via tryLoadConfigurationFileForProjectAsync 1`] = `
Object {
  "error": "",
  "log": "",
  "verbose": "Config file \\"<project root>/src/test/project-referencing-rig/config/simplestConfigFile.json\\" does not exist. Attempting to load via rig.[n]",
  "warning": "",
}
`;

exports[`ConfigurationFile loading a rig throws an error when a config file doesn't exist in a project referencing a rig, which also doesn't have the file 1`] = `[Error: File does not exist: <project root>/src/test/project-referencing-rig/config/notExist.json]`;

exports[`ConfigurationFile loading a rig throws an error when a config file doesn't exist in a project referencing a rig, which also doesn't have the file 2`] = `
Object {
  "error": "",
  "log": "",
  "verbose": "Configuration file \\"<project root>/src/test/project-referencing-rig/config/notExist.json\\" not found.[n]Configuration file \\"<project root>/src/test/project-referencing-rig/node_modules/test-rig/profiles/default/config/notExist.json\\" not found.[n]Configuration file \\"config/notExist.json\\" not found in rig (\\"<project root>/src/test/project-referencing-rig/node_modules/test-rig/profiles/default\\")[n]",
  "warning": "",
}
`;
